hydrolib/core/io/ini/serializer.py

```diff
@@ -0,0 +1,246 @@
+    max_key_length: int
+    max_value_length: int
+    max_datablock_element_length: Optional[Sequence[int]] = None
+
+    @staticmethod
+    def _calculate_max_length_datablock(
+        datablock: Optional[Sequence[Sequence[str]]],
+    ) -> Optional[Sequence[int]]:
+        if datablock is None or len(datablock) < 1:
+            return None
+
+        def folder(acc: Tuple, row: Sequence[str]) -> Tuple:
+            return tuple(max(curr_max, len(elem)) for curr_max, elem in zip(acc, row))
+
+        return reduce(folder, datablock, (0,) * len(datablock[0]))
+
+    @classmethod
+    def from_section(cls, section: Section):
+        properties = list(
+            prop for prop in section.content if isinstance(prop, Property)
+        )
+
+        if len(properties) > 0:
+            max_key_length = max(len(prop.key) for prop in properties)
+            max_value_length = max(
+                len(prop.value) if prop.value is not None else 0 for prop in properties
+            )
+        else:
+            max_key_length = 0
+            max_value_length = 0
+
+        return cls(
+            max_key_length=max_key_length,
+            max_value_length=max_value_length,
+            max_datablock_element_length=_Lengths._calculate_max_length_datablock(
+                section.datablock
+            ),
+        )
+
+
+class Serializer:
+    """Serializer serializes Document to its corresponding lines."""
+
+    def __init__(self, config: SerializerConfig):
+        """Creates a new Serializer with the provided configuration.
+
+        Args:
+            config (SerializerConfig): The configuration of this Serializer.
+        """
+        self._config = config
+
+    @staticmethod
+    def _interweave(iterable: Iterable, val: Any) -> Iterable:
+        # Interweave the provided iterable with the provided value:
+        # iterable_element, val, iterable_element, val, ...
+
+        # Note that this will interweave with val without making copies
+        # as such it is the same object being interweaved.
+        return chain.from_iterable(zip(iterable, repeat(val)))
+
+    def _serialize_comment_block(
+        self, block: CommentBlock, indent_size: int = 0
+    ) -> Iterable[str]:
+        indent = " " * indent_size
+        return (f"{indent}{self._config.comment_delimiter} {l}" for l in block.lines)
+
+    def _serialize_comment_header(
+        self, header_blocks: Iterable[CommentBlock]
+    ) -> Iterable[str]:
+        blocks_as_lines = (
+            self._serialize_comment_block(block) for block in header_blocks
+        )
+        # We separate comment blocks in the header with an empty line.
+        blocks_as_lines = Serializer._interweave(blocks_as_lines, [""])  # type: ignore
+
+        return chain.from_iterable(blocks_as_lines)
+
+    def _serialize_section_header(self, section_header: str) -> Iterable[str]:
+        indent = " " * (self._config.section_indent)
+        return [f"{indent}[{section_header}]"]
```

Comment:
It's a little nitpicky but I'd personally return the value not as a list. To keep the scope of the function not bother with how it's used later on. But feel free to ignore this, it's a private function anyway.

Reply:
The reason I wanted to return an `Iterable` is because it allows me to just use a `chain` call in the calling level. You are right however that a list might not make so much sense. I have replaced the return list with a yield, which works just as well (both here and at a similar return statement in the Serializer).